File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        vector<vector<int>> dp(matrix);
        
        for(int i=1; i<dp.size(); i++){
            for(int j=0; j<dp.size(); j++){
                dp[i][j] += min({dp[i-1][j], dp[i-1][max(0, j-1)], dp[i-1][min((int)dp.size()-1, j+1)]});
            }
        }

        return *min_element(dp[dp.size()-1].begin(), dp[dp.size()-1].end());
    }
};

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        @lru_cache(None)
        def dp(i, j):
            if j < 0 or j >= n: return float('inf')
            if i==n-1: return matrix[i][j]
            return min(dp(i+1, j), dp(i+1, j-1), dp(i+1, j+1)) + matrix[i][j]
        return min(dp(0, j) for j in range(n))
